apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryParameterCollection.Data.cs

```diff
@@ -26,105 +27,109 @@ public sealed partial class BigQueryParameterCollection : DbParameterCollection
+                "The argument must not be null and be of type {0}", nameof(BigQueryParameter));
+            _parameters.Add(bigQueryParameter);
+            return _parameters.Count - 1;
         }
 
         /// <inheritdoc />
-        public override bool Contains(object value)
-        {
-            throw new NotImplementedException();
-        }
+        public override bool Contains(object value) => _parameters.Contains(value as BigQueryParameter);
 
         /// <inheritdoc />
-        public override bool Contains(string value)
-        {
-            throw new NotImplementedException();
-        }
+        public override bool Contains(string value) => _parameters.Any(x => x.ParameterName.Equals(value));
 
         /// <inheritdoc />
-        public override int IndexOf(string parameterName)
-        {
-            throw new NotImplementedException();
-        }
+        public override int IndexOf(string parameterName) => _parameters.FindIndex(x => x.ParameterName.Equals(parameterName));
 
         /// <inheritdoc />
-        public override void RemoveAt(string parameterName)
-        {
-            throw new NotImplementedException();
-        }
+        public override void RemoveAt(string parameterName) => _parameters.RemoveAt(IndexOf(parameterName));
 
         /// <inheritdoc />
         protected override void SetParameter(int index, DbParameter value)
-        {
-            throw new NotImplementedException();
-        }
+            => _parameters[index] = (BigQueryParameter)value;
 
         /// <inheritdoc />
         protected override void SetParameter(string parameterName, DbParameter value)
         {
-            throw new NotImplementedException();
+            int index = IndexOf(parameterName);
+            if (index == -1)
+            {
+                _parameters.Add((BigQueryParameter)value);
+            }
+            else
+            {
+                _parameters[index] = (BigQueryParameter)value;
+            }
         }
 
 #if NET45
         /// <inheritdoc />
-        public override bool IsFixedSize { get; }
+        public override bool IsFixedSize => false;
         /// <inheritdoc />
-        public override bool IsReadOnly { get; }
+        public override bool IsReadOnly => false;
         /// <inheritdoc />
-        public override bool IsSynchronized { get; }
+        public override bool IsSynchronized => false;
 #endif
         /// <inheritdoc />
-        public override object SyncRoot { get; }
+        public override object SyncRoot => _parameters;
 
         /// <inheritdoc />
         public override void Insert(int index, object value)
-        {
-            throw new NotImplementedException();
-        }
+            => _parameters.Insert(index, (BigQueryParameter)value);
 
         /// <inheritdoc />
         public override void Remove(object value)
-        {
-            throw new NotImplementedException();
-        }
+            => _parameters.Remove((BigQueryParameter)value);
```

Comment:
Do we want this to throw an exception if the value isn't a BigQueryParameter? I'd expect it to basically ignore any value not in the collection.

Reply:
it should throw -- that's a code error on the part of the developer.  He should not attempt to add or remove any type that isnt a BQParameter.  Ignoring the issue on Remove might create false expectations on Add -- where we cannot ignore the error.